interfaces: add audio-playback/record and pulseaudio spread tests

jdstrand opened the pull request:
This builds on https://github.com/snapcore/snapd/pull/5644 and adds spread tests for pulseaudio, audio-playback and audio-record.

bboozzoo commented on interfaces/builtin/audio_playback.go:
```diff
@@ -0,0 +1,179 @@
+
+# Shared memory based communication with clients
+/{run,dev}/shm/pulse-shm-* mrwk,
+
```
The `pulseaudio` interface also has `/usr/share/applications/ r,` in this place. AsI understand it's a pulseaudio specific entry and may not fit well into the generic `audio-playback` interface.

bboozzoo commented on interfaces/builtin/audio_playback.go:
```diff
@@ -0,0 +1,179 @@
+/sys/devices/virtual/dmi/id/sys_vendor r,
+/sys/devices/virtual/dmi/id/bios_vendor r,
+/sys/**/sound/** r,
+
```
The pulseaudio iface additionally carries this entry:
```
# FIXME: use udev queries to make this more specific
/run/udev/data/** r,
```

bboozzoo commented on interfaces/builtin/audio_playback.go:
```diff
@@ -0,0 +1,179 @@
+network netlink raw,
+/sys/devices/virtual/dmi/id/sys_vendor r,
+/sys/devices/virtual/dmi/id/bios_vendor r,
+/sys/**/sound/** r,
```
I understand this replaces `/dev/snd/*` rules?

anonymouse64 commented:
The diff here is confusing, as it says that `interfaces/builtin/audio_playback.go` (and some other files) is added as a new file in this PR, however it seems that it was already added with the merge of 5644 with https://github.com/snapcore/snapd/commit/76492e4f1d334c890f403bc45c10573400413616 ? Indeed, audio_playback.go is already present in master at https://github.com/snapcore/snapd/blob/master/interfaces/builtin/audio_playback.go.

jdstrand commented:
@anonymouse64 - yeah, for some reason this didn't git updated when I merged PR #5644. I updated this accordingly. @bboozzoo - can you take another look? (what you reviewed was already approved and merged).

anonymouse64 commented:
Thanks @jdstrand the diff looks good now

jdstrand commented:
> There seems to be a snap file committed at `tests/lib/snaps/test-snapd-audio-record/test-snapd-audio-record_1_amd64.snap` in this PR, I think it should probably be removed.

Fixed and force pushed for clarity. That is the last force push I'll do on this PR.

zyga commented on tests/lib/snaps/test-snapd-audio-record/Makefile:
```diff
@@ -0,0 +1,8 @@
+all:
+	cd ./src && make all
```
The usual way to handle this is:
```suggestion
all install clean:
	$(MAKE) -C src $@
```

zyga commented on tests/main/interfaces-audio-playback-record/task.yaml:
```diff
@@ -0,0 +1,108 @@
+summary: Ensure that the audio-playback/record interface works
+
+# Only test on classic Ubuntu amd64 systems
```
Rather than saying what the subsequent line does, it would be better to say why. I suspect it is because we don't have the appropriate patches anywhere else but I don't know.

zyga commented on tests/lib/snaps/test-snapd-pulseaudio/Makefile:
```diff
@@ -0,0 +1,8 @@
+all:
```
Same comment about makefile as above

zyga commented:
@jdstrand this needs to be re-visited.

jdstrand commented on tests/main/interfaces-pulseaudio/task.yaml:
```diff
@@ -0,0 +1,82 @@
+    apt autoremove --purge -y pulseaudio
+
+execute: |
+    trap 'HOME=/home/test XDG_RUNTIME_DIR=/run/user/12345 su -p -c "pulseaudio --kill" test || true' EXIT
```
@zyga - I recall trying this in restore and it hung the spread test but this use of trap worked.

jdstrand commented:
@zyga - I believe I addressed your points. I didn't use 'wait' since pulseaudio has an --exit-idle-time argument and I moved the --kill to restore. For some reason I had to force push even though I pulled from master and fetch/merged again to make everything in sync.

jdstrand commented:
I have to fix this for earlier releases. Marking as blocked for now.

jdstrand commented:
> I have to fix this for earlier releases. Marking as blocked for now.

The issue was using 'apt autoremove --purge' is unsupported on trusty and not using 'recsimple' with the test-snapd-audio-playback-record tests like we do with pulseaudio. Removing Blocked tag, but will keep an eye on this if tests don't go green now.

pedronis commented:
@jdstrand should this be targeted for 2.43 (it might be hard) ?

jdstrand commented:
@pedronis - all this needs is for @zyga to review. Yes, 2.43 seems fine to me.

zyga commented on tests/lib/snaps/test-snapd-audio-record/files/bin/pawrap:
```diff
@@ -0,0 +1,16 @@
+
+papath="$XDG_RUNTIME_DIR/../pulse/native"
+if [ ! -e "$papath" ]; then
+    echo "Could not find '$papath'"
```
```suggestion
    echo "cannot find '$papath'"
```

zyga commented on tests/lib/snaps/test-snapd-pulseaudio/files/bin/pawrap:
```diff
@@ -0,0 +1,16 @@
+
+papath="$XDG_RUNTIME_DIR/../pulse/native"
+if [ ! -e "$papath" ]; then
+    echo "Could not find '$papath'"
```
```suggestion
    echo "cannot find '$papath'"
```

zyga commented on tests/lib/snaps/test-snapd-audio-record/snapcraft.yaml:
```diff
@@ -0,0 +1,81 @@
+environment:
+  LD_LIBRARY_PATH: $SNAP/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/pulseaudio
+
+apps:
```
I must say that the plug allocations to the apps are really neat here.

zyga commented on tests/main/interfaces-audio-playback-record/task.yaml:
```diff
@@ -0,0 +1,112 @@
+
+    echo "Create XDG_RUNTIME_DIR=/run/user/12345"
+    # shellcheck disable=SC2174
+    mkdir -m 700 -p /run/user/12345 || true
```
Do you expect this to fail?

jdstrand replied:
No, but if it happens to exist I was thinking that was non-fatal.

pedronis replied:
-p should take care of that already afaik

zyga commented on tests/main/interfaces-audio-playback-record/task.yaml:
```diff
@@ -0,0 +1,112 @@
+
+    echo "Then wait for the socket to show up"
+    count=0
+    while sleep 1 && [ ! -S /run/user/12345/pulse/native ]; do
```
I think this whole paragraph can be reduced to:
```
retry-tool -n 10 test -S /run/user/12345/pulse/native
```

zyga commented on tests/main/interfaces-audio-playback-record/task.yaml:
```diff
@@ -0,0 +1,112 @@
+
+    echo "Check pulseaudio"
+    count=0
+    while sleep 1 && ! HOME=/home/test XDG_RUNTIME_DIR=/run/user/12345 su -p -c "pulseaudio --check" test ; do
```
I think you can also reduce this with retry-tool.

zyga commented on tests/main/interfaces-audio-playback-record/task.yaml:
```diff
@@ -0,0 +1,112 @@
+environment:
+    PLAY_FILE: "/snap/test-snapd-audio-record/current/usr/share/sounds/alsa/Noise.wav"
+    # today, only 19.04 and 19.10 have a mediating pulseaudio
+    EXFAIL: "ubuntu-14 ubuntu-16 ubuntu-18"
```
Nitpick: based on how this is used I think it would be simpler if it was the full regular expression matching the expected failures: `"ubuntu-1[468]"`. This would remove the whitespace trickiness, the loop and everything below.

jdstrand replied:
Done

zyga commented on tests/main/interfaces-audio-playback-record/task.yaml:
```diff
@@ -0,0 +1,112 @@
+    snap disconnect test-snapd-audio-record:audio-record
+
+    mediating_pa="yes"
+    for regex in $EXFAIL ; do
```
Some ideas on how to simplify this above next to the variable definition.

zyga commented on tests/main/interfaces-audio-playback-record/task.yaml:
```diff
@@ -0,0 +1,112 @@
+restore: |
+    HOME=/home/test XDG_RUNTIME_DIR=/run/user/12345 su -p -c "pulseaudio --kill" test || true
+    snap remove test-snapd-audio-record
+    apt-get autoremove --purge -y pulseaudio pulseaudio-utils
```
Do we need `pulseaudio-utils`?

jdstrand replied:
Yes, for paplay in "The unconfined user can play audio"

zyga commented on tests/main/interfaces-pulseaudio/task.yaml:
```diff
@@ -0,0 +1,87 @@
+
+    echo "Then wait for the socket to show up"
+    count=0
+    while sleep 1 && [ ! -S /run/user/12345/pulse/native ]; do
```
Another place where we could use `retry-tool`

zyga commented on tests/main/interfaces-pulseaudio/task.yaml:
```diff
@@ -0,0 +1,87 @@
+
+    echo "Check pulseaudio"
+    count=0
+    while sleep 1 && ! HOME=/home/test XDG_RUNTIME_DIR=/run/user/12345 su -p -c "pulseaudio --check" test ; do
```
And here.

jdstrand commented on tests/lib/snaps/test-snapd-audio-record/src/Makefile:
```diff
@@ -0,0 +1,20 @@
+CFLAGS += -g -O0 -Wall -Wstrict-prototypes
```
For test code, I tend to not like to optimize in case there is weird behavior since it takes the compiler out of the question.

jdstrand commented:
@anonymouse64 and @bboozzoo - one of you looked at this before; does one of you want to perform the second review?

bboozzoo commented on tests/main/interfaces-audio-playback-record/task.yaml:
```diff
@@ -0,0 +1,104 @@
+summary: Ensure that the audio-playback/record interface works
+
+# Only classic Ubuntu has the pulseaudio mediation patches
+systems: [ ubuntu-1*-*64 ]
```
The test will stop running on 20.04?

jdstrand replied:
Done

anonymouse64 commented on tests/main/interfaces-audio-playback-record/task.yaml:
```diff
@@ -0,0 +1,104 @@
+    HOME=/home/test XDG_RUNTIME_DIR=/run/user/12345 su -p -c "/snap/test-snapd-audio-record/current/bin/parec-simple" test
+
+    echo "The audio-playback interface is connected by default"
+    snap interfaces -i audio-playback | MATCH ":audio-playback .*test-snapd-audio-record"
```
```
'snap interfaces' is deprecated; use 'snap connections'.
```
Can we use `snap connections` here instead?

anonymouse64 commented on tests/main/interfaces-audio-playback-record/task.yaml:
```diff
@@ -0,0 +1,104 @@
+    snap interfaces -i audio-playback | MATCH ":audio-playback .*test-snapd-audio-record"
+
+    echo "The audio-record interface is disconnected by default"
+    snap interfaces -i audio-record | MATCH "\- +test-snapd-audio-record:audio-record"
```
ditto for `snap connections`

anonymouse64 commented on tests/main/interfaces-pulseaudio/task.yaml:
```diff
@@ -0,0 +1,82 @@
+    HOME=/home/test XDG_RUNTIME_DIR=/run/user/12345 su -p -c "/snap/test-snapd-pulseaudio/current/bin/parec-simple" test
+
+    echo "The interface is disconnected by default"
+    snap interfaces -i pulseaudio | MATCH "\- +test-snapd-pulseaudio:pulseaudio"
```
ditto for `snap connections`

anonymouse64 commented on tests/main/interfaces-audio-playback-record/task.yaml:
```diff
@@ -0,0 +1,104 @@
+    snap connect test-snapd-audio-record:audio-record
+
+    echo "Then the snap can record audio"
+    su -l -c "test-snapd-audio-record.recsimple" test
```
should we add a test before audio-record is connected that shows there are apparmor denials when the plug is not connected and then check that the app doesn't have any apparmor denials when the plug is connected?

jdstrand replied:
> should we add a test before audio-record is connected that shows there are apparmor denials when the plug is not connected and then check that the app doesn't have any apparmor denials when the plug is connected?

This is not something we do in other interface tests, largely because kernel rate limiting is problematic. We've historically only checked the access and if the test (positive or negative) fails, we see in the logs the audit messages.

jdstrand commented on tests/lib/snaps/test-snapd-pulseaudio/files/bin/pawrap:
```diff
@@ -0,0 +1,16 @@
+#!/bin/sh
```
Well, not without rearranging the tests under a single snap. I wanted different snaps since the pulseaudio test will always be just pulseaudio, but audio-playback is expected to be extended to include pipewire, etc. Merging everything into a shared 'test-snapd-audio' snap is possible, but messier for this reason IMO. Furthermore, since these require compiling code, they are built out of bzr branches in LP so I can't do symlinks/hardlinks/similar since they are distinct trees in LP and the code here is merely a mirror of that code (I'm not entirely sure why this is the current methodology, but I'm told this is the way to do it)).

anonymouse64 replied:
this is fine for now, we discussed having a test-snapd repo for all of these snaps that makes sure the snaps are updated and such and have that repo be independent from snapd so I guess we can leave this here for now until we do that switch.

jdstrand commented on tests/lib/snaps/test-snapd-audio-record/src/Makefile:
```diff
@@ -0,0 +1,20 @@
+CFLAGS += -g -O0 -Wall -Wstrict-prototypes
+
+# snapcraft will copy anything from here
+INSTALL_DIR := ../../install/bin
```
Ok, I updated the makefile to use DESTDIR instead, tested locally and then pushed to bzr and tested the built snaps. I then pushed the changes to this PR (ie after the snaps were built and available in edge) so spread will rerun with these updated snaps.

jdstrand commented on tests/lib/snaps/test-snapd-pulseaudio/src/Makefile:
```diff
@@ -0,0 +1,20 @@
+CFLAGS += -g -O0 -Wall -Wstrict-prototypes
+
+# snapcraft will copy anything from here
+INSTALL_DIR := ../../install/bin
```
Done
